rcsb/workflow/wuw/PdbCsmImageWorkflow.py

```diff
@@ -0,0 +1,157 @@
+                model_list.append(f"{model_id} {model_path} computational")
+    return model_list
+
+
+def images_gen_lists(args: dict) -> None:
+    """Generate lists of pdbs/csms in files."""
+    pdb_id_list = get_pdb_list(pdb_gz_path=args['pdb_gz_path'], update_all_images=args['update_all_images'], pdb_base_dir=args['pdb_base_dir'])
+    comp_id_list = [] if args["imgs_exclude_models"] else get_csm_list(csm_gz_path=args['csm_gz_path'], update_all_images=['update_all_images'], csm_base_dir=args['csm_base_dir'])
+
+    # Print results, combine, and shuffle
+    if len(pdb_id_list) < 1 and len(comp_id_list) < 1:
+        msg = "pdb and csm id list empty"
+        raise ValueError(msg)
+    msg = f"There are {len(pdb_id_list)} pdb_ids and {len(comp_id_list)} comp_ids for which to generate BCIF and images"
+
+    full_id_list = pdb_id_list + comp_id_list
+    random.shuffle(full_id_list)
+
+    steps = int(len(full_id_list) / int(args["num_workers"]) )
+    for i in range(0, len(full_id_list), steps):
+        Path(args["id_list_path"]).mkdir(parents=True, exist_ok=True)
+        with Path.open(args["id_list_path"] + str(int(i/steps)), "w") as f:
+            for line in full_id_list[i : i + steps]:
+                f.write(line + "\n")
+
+def images_gen_jpgs(args: dict) -> None:
+    """Generate jpgs for given pdb/csm list."""
+    with Path.open(args["id_list_path"] + args['file_number'], 'r') as f:
+        id_list = [line.rstrip('\n') for line in f]
+    if not isinstance(id_list, list):
+        msg = "id_list not a list"
+        raise TypeError(msg)
+    
+    for line in id_list:
+        # Requirements:
+        # 1. bcif files must be unzipped
+        # 2. bcif files must be in a local dir (otherwise I'll have to add a curl step)
+        # 3. bcif files do not need conversion from cif files. This should be taken care of in bcif workflow
+        # 
+        # todo: add a step that checks:
+        # 1. the dir exists
+        # 2. the file is a bcif with data inside of it
+        # 3. the output dir is availible
+
+        file_id, bcif_file_name, sdm = line.split(" ")
+        content_type_dir = "pdb/" if sdm == "experimental" else "csm/"
+
+        bcif_file_path = args["pdb_base_dir"] + bcif_file_name if sdm == "experimental" else args["csm_base_dir"] + bcif_file_name 
+        
+        out_path = args["jpgs_out_dir"] + content_type_dir
+        Path(out_path).mkdir(parents=True, exist_ok=True)
+        
+        ### run_molrender ###
+        cmd = [
+            args['jpg_xvfb_executable'],
+            '-a',
+            '-s', f'-ac -screen 0 {args['jpg_screen']}',
+            args["molrender_exe"],
+            'all',
+            bcif_file_path,
+            out_path,
+            '--height', args['jpg_height'],
+            '--width', args['jpg_width'],
+            '--format', args['jpg_format'],
+            ]
+        if args['jpg_additional_cmds'] is None:
+            cmd = [*cmd, args['jpg_additional_cmds']]
```

Comment:
We're going to have to think about how we'd enable testing of this (i.e., how to install the molrender software into Azure, or GitHub Actions when we get there).

Reply:
So this task is particularly weird... the rcsb-wuw task will build a shell command to be sumitted to Nersc... Done. The shell command is a script that builds a new shell command for each item in a file then runs that command. Why not just combine these shell-building-commands into one? Because the second one needs to open and read the file containing the pdbs and csm to do. Currently, this file is located at Nersc, so they can't be combined. The way around this would be to submit the task with the specific pdb or csm to do instead of writing a list of them to a file. Not writing to a file has the problem of rerunning things if they fail. Yes we can with airflow, but things can still break easier this way. This file version is much simpler as well.